TypeCobol.LanguageServer/WorkspaceProjectStore.cs

```diff
@@ -0,0 +1,347 @@
+        /// Refresh all documents in all WorkspaceProject instances
+        /// </summary>
+        /// <param name="workspace">The main Workspace instance</param>
+        internal void RefreshOpenedFiles(Workspace workspace)
+        {
+            foreach (var wksProject in _workspaceProjects.Values)
+            {
+                wksProject.RefreshOpenedFiles(workspace);
+            }
+        }
+
+        /// <summary>
+        /// Move a Document from one WorkspaceProject to another WorkSpaceProject.
+        /// </summary>
+        /// <param name="docContext">The Document Context instance to move</param>
+        /// <param name="fromWorkspaceProject">The Source WorkspaceProject instance</param>
+        /// <param name="toWorkspaceProject">The Target WorkspaceProject instance.</param>
+        private void MoveWorkspaceProjectDocument(DocumentContext docContext, WorkspaceProject fromWorkspaceProject, WorkspaceProject toWorkspaceProject)
+        {
+            System.Diagnostics.Debug.Assert(docContext != null && fromWorkspaceProject != null && toWorkspaceProject != null);
+            System.Diagnostics.Debug.Assert(fromWorkspaceProject.Contains(docContext));
+            System.Diagnostics.Debug.Assert(!toWorkspaceProject.Contains(docContext));
+
+            fromWorkspaceProject.RemoveDocument(docContext);
+            toWorkspaceProject.AddDocument(docContext);            
+        }
+
+        /// <summary>
+        /// Add the given Document context to the Workspace Project corresponding to the given project's key.
+        /// The project's key must correspond to an existing WorkspaceProjet, if the projectKey value is null then 
+        /// the Default Workspace Project will be used.
+        /// </summary>
+        /// <param name="docContext">The DocumentContext to be added</param>
+        /// <param name="projectKey">The target Project's key</param>
+        /// <exception cref="UnknownProjectException">If the project key does correspond to an existing project</exception>
+        /// <returns>true if the document is added, false if it already exists in the target project</returns>
+        internal bool AddDocument(DocumentContext docContext, string projectKey)
+        {
+            WorkspaceProject workspaceProject = FindProject(projectKey);
+            if(workspaceProject == null)
+                throw new UnknownProjectException(projectKey);
+            if (workspaceProject.AddDocument(docContext))
+            {   // Add the document to all opened document set.
+                return this._allOpenedDocuments.TryAdd(docContext.Uri, docContext);
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Rmove a document from it's uri
+        /// </summary>
+        /// <param name="docUri">The document's uri</param>
+        /// <returns>The removed document if any, null if no document was removed</returns>
+        internal DocumentContext RemoveDocumentFromUri(Uri docUri)
```

Comment:
This is more a Close method than a Remove. I mean WorkspaceProjectStore _owns_ documents so we open/close them. A WorkspaceProject is associated to some documents but does not own them so Add/Remove is correct.